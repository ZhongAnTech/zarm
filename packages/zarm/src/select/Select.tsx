import React, { PureComponent } from 'react';
import classnames from 'classnames';
import isEqual from 'lodash/isEqual';
import BaseSelectProps from './PropsType';
import Picker from '../picker';
import parseProps from '../picker-view/utils/parseProps';
import removeFnFromProps from '../picker-view/utils/removeFnFromProps';
import { WheelItem } from '../wheel/PropsType';

export interface SelectProps extends BaseSelectProps {
  prefixCls?: string;
  className?: string;
}

export interface SelectState {
  selectValue: Array<WheelItem>;
  visible: boolean;
}

export default class Select extends PureComponent<SelectProps, SelectState> {
  static defaultProps = {
    prefixCls: 'za-select',
    dataSource: [],
    valueMember: 'value',
    itemRender: (data) => data && data.label,
    cols: Infinity,
    hasArrow: true,
    maskClosable: true,
    displayRender: (selected) => selected.map((item) => item && item.label),
    onClick: () => {},
  };

  constructor(props) {
    super(props);
    this.state = {
      visible: false,
      selectValue:
        parseProps.isValueValid(props.defaultValue || props.value) &&
        parseProps.getSource(props).objValue,
    };
  }

  static getDerivedStateFromProps(props, state) {
    if (
      !isEqual(
        removeFnFromProps(props, ['onOk', 'onCancel', 'onChange']),
        removeFnFromProps(state.prevProps, ['onOk', 'onCancel', 'onChange']),
      )
    ) {
      return {
        prevProps: props,
        selectValue:
          parseProps.isValueValid(props.defaultValue || props.value) &&
          parseProps.getSource(props).objValue,
      };
    }

    return null;
  }

  handleClick = () => {
    const { disabled } = this.props;
    if (disabled) {
      return false;
    }
    this.setState({
      visible: true,
    });
  };

  onChange = (selected) => {
    const { onChange } = this.props;
    if (typeof onChange === 'function') {
      onChange(selected);
    }
  };

  onOk = (selected) => {
    const { onOk } = this.props;
    this.setState(
      {
        selectValue: selected,
        visible: false,
      },
      () => {
        if (typeof onOk === 'function') {
          onOk(selected);
        }
      },
    );
  };

  // 点击取消
  onCancel = () => {
    const { onCancel } = this.props;
    this.setState({
      visible: false,
    });

    if (typeof onCancel === 'function') {
      onCancel();
    }
  };

  render() {
    const {
      prefixCls,
      placeholder,
      className,
      disabled,
      displayRender,
      locale,
      value,
      hasArrow,
      ...others
    } = this.props;
    const { visible, selectValue = [] } = this.state;
    const cls = classnames(prefixCls, {
      [`${prefixCls}--placeholder`]: !selectValue.length,
      [`${prefixCls}--disabled`]: disabled,
      [`${prefixCls}--arrow`]: hasArrow,
    });
    return (
      <div className={cls} onClick={this.handleClick}>
        <div className={`${prefixCls}__input`}>
          <div className={`${prefixCls}__value`}>
            {(selectValue.length && displayRender!(selectValue || [])) ||
              placeholder ||
              locale!.placeholder}
          </div>
        </div>
        <Picker
          {...others}
          className={className}
          visible={visible}
          value={value}
          onOk={this.onOk}
          onChange={this.onChange}
          onCancel={this.onCancel}
        />
      </div>
    );
  }
}
