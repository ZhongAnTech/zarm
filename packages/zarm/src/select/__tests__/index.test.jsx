import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Select from '../index';

describe('Select', () => {
  const fakeTimers = () => {
    performance.timing = () => {};
  };
  fakeTimers();

  it('Select', () => {
    const wrapper = mount(
      <Select
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Select disabled', () => {
    const wrapper = mount(
      <Select
        disabeld
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
      />,
    );

    wrapper.find('.za-select').simulate('click');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('render defaultValue correctly ', () => {
    const wrapper = mount(
      <Select
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
        defaultValue="2"
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Cascader Select', () => {
    // jest.useFakeTimers();
    const wrapper = mount(
      <Select
        dataSource={[
          {
            value: '1',
            label: '选项一',
            children: [
              { value: '11', label: '选项一' },
              { value: '12', label: '选项二' },
            ],
          },
          {
            value: '2',
            label: '选项一',
            children: [
              { value: '21', label: '选项一' },
              { value: '22', label: '选项二' },
            ],
          },
        ]}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.setProps({ value: ['1', '12'] });
    // jest.runAllTimers();
    wrapper.unmount();
  });

  it('Cascader Select init value', () => {
    // jest.useFakeTimers();
    const wrapper = mount(
      <Select
        dataSource={[
          {
            value: '1',
            label: '选项一',
            children: [
              { value: '11', label: '选项一' },
              { value: '12', label: '选项二' },
            ],
          },
          {
            value: '2',
            label: '选项一',
            children: [
              { value: '21', label: '选项一' },
              { value: '22', label: '选项二' },
            ],
          },
        ]}
        value={['1', '12']}
        displayAddon="-"
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('receive new dataSource', () => {
    const wrapper = shallow(
      <Select
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
      />,
    );
    wrapper.setProps({
      dataSource: [
        { value: 'a', label: '选项一' },
        { value: 'b', label: '选项二' },
        { value: 'c', label: '选项三' },
      ],
    });
  });

  it('receive new value', () => {
    const wrapper = shallow(
      <Select
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
      />,
    );
    wrapper.setProps({ value: '1' });
  });

  it('receive new cascader dataSource', () => {
    const wrapper = shallow(
      <Select
        dataSource={[
          {
            value: '1',
            label: '选项一',
            children: [
              { value: '11', label: '选项一' },
              { value: '12', label: '选项二' },
            ],
          },
          {
            value: '2',
            label: '选项一',
            children: [
              { value: '21', label: '选项一' },
              { value: '22', label: '选项二' },
            ],
          },
        ]}
      />,
    );

    wrapper.setProps({
      dataSource: [
        {
          value: '3',
          label: '选项一',
          children: [
            { value: '31', label: '选项一' },
            { value: '32', label: '选项二' },
          ],
        },
        {
          value: '4',
          label: '选项一',
          children: [
            { value: '41', label: '选项一' },
            { value: '42', label: '选项二' },
          ],
        },
      ],
    });
  });

  it('should trigger onOk when press ok button', () => {
    const onOkFn = jest.fn();
    const onCancelFn = jest.fn();

    const wrapper = mount(
      <Select
        dataSource={[
          {
            value: '1',
            label: '选项一',
            children: [
              { value: '11', label: '选项一' },
              { value: '12', label: '选项二' },
            ],
          },
          {
            value: '2',
            label: '选项一',
            children: [
              { value: '21', label: '选项一' },
              { value: '22', label: '选项二' },
            ],
          },
        ]}
        value={['1', '12']}
        onOk={onOkFn}
        onCancel={onCancelFn}
      />,
    );

    wrapper.find('.za-select').simulate('click');
    jest.useFakeTimers();
    wrapper.find('.za-picker__submit').simulate('click');
    jest.runAllTimers();
    expect(onOkFn).toBeCalled();
    expect(onCancelFn).not.toBeCalled();
  });

  it('should trigger onCancel when press cancel button', () => {
    const onOkFn = jest.fn();
    const onCancelFn = jest.fn();

    const wrapper = mount(
      <Select
        dataSource={[
          {
            value: '1',
            label: '选项一',
            children: [
              { value: '11', label: '选项一' },
              { value: '12', label: '选项二' },
            ],
          },
          {
            value: '2',
            label: '选项一',
            children: [
              { value: '21', label: '选项一' },
              { value: '22', label: '选项二' },
            ],
          },
        ]}
        defaultValue={['1', '12']}
        onOk={onOkFn}
        onCancel={onCancelFn}
      />,
    );
    wrapper.find('.za-select').simulate('click');
    jest.useFakeTimers();
    wrapper.find('.za-picker__cancel').simulate('click');
    expect(onCancelFn).toBeCalled();
    expect(onOkFn).not.toBeCalled();
  });

  // it('should trigger onMaskClick when click mask', () => {
  //   const onOkFn = jest.fn();
  //   const onMaskClick = jest.fn();

  //   const wrapper = mount(
  //     <Select
  //       dataSource={[
  //         {
  //           value: '1',
  //           label: '选项一',
  //           children: [
  //             { value: '11', label: '选项一' },
  //             { value: '12', label: '选项二' },
  //           ],
  //         },
  //         {
  //           value: '2',
  //           label: '选项一',
  //           children: [
  //             { value: '21', label: '选项一' },
  //             { value: '22', label: '选项二' },
  //           ],
  //         },
  //       ]}
  //       visible
  //       defaultValue={['1', '12']}
  //       onOk={onOkFn}
  //       onMaskClick={onMaskClick}
  //     />
  //   );

  //   wrapper.find('.za-mask').simulate('click');
  //   expect(onMaskClick).toBeCalled();
  // });
});
