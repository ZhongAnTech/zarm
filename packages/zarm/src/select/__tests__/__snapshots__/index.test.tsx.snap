// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Cascader Select 1`] = `
<div>
  <div
    class="za-select"
  >
    <div
      class="za-select__input"
    >
      <div
        class="za-select__value"
      >
        选项一
        选项二
      </div>
    </div>
    <div
      class="za-select__arrow"
    />
  </div>
</div>
`;

exports[`Select Select 1`] = `
<div>
  <div
    class="za-select"
  >
    <div
      class="za-select__input"
    >
      <div
        class="za-select__value"
      >
        选项一
      </div>
    </div>
    <div
      class="za-select__arrow"
    />
  </div>
</div>
`;

exports[`Select Select disabled 1`] = `
<div>
  <div
    class="za-select za-select--disabled"
  >
    <div
      class="za-select__input"
    >
      <div
        class="za-select__value"
      >
        选项一
      </div>
    </div>
    <div
      class="za-select__arrow"
    />
  </div>
</div>
`;

exports[`Select render defaultValue correctly  1`] = `
<div>
  <div
    class="za-select"
  >
    <div
      class="za-select__input"
    >
      <div
        class="za-select__value"
      >
        选项二
      </div>
    </div>
    <div
      class="za-select__arrow"
    />
  </div>
</div>
`;
